File: functions/chatbot.js
// functions/chatbot.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

// Configurar __dirname para módulos ES
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Função handler para Netlify Functions
const handler = async function(event, context) {
  // Adicionar headers CORS para todas as respostas
  const headers = {
    'Content-Type': 'application/json',
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Headers': 'Content-Type',
    'Access-Control-Allow-Methods': 'GET, POST, OPTIONS'
  };
  
  // Lidar com requisições OPTIONS (preflight CORS)
  if (event.httpMethod === 'OPTIONS') {
    return {
      statusCode: 200,
      headers,
      body: JSON.stringify({ message: 'CORS preflight successful' })
    };
  }

  try {
    // Verificar se é uma requisição POST
    if (event.httpMethod !== 'POST') {
      return {
        statusCode: 405,
        headers,
        body: JSON.stringify({ error: 'Método não permitido' })
      };
    }

    // Obter a pergunta do corpo da requisição
    const { question } = JSON.parse(event.body);
    
    if (!question) {
      return {
        statusCode: 400,
        headers,
        body: JSON.stringify({ error: 'Pergunta não fornecida' })
      };
    }

    console.log('Processando pergunta:', question);

    // Carregar dados de eventos e notícias
    let events = [];
    let news = [];

    try {
      // Caminho para os arquivos JSON
      const eventsPath = path.join(__dirname, '../public/events_data.json');
      const newsPath = path.join(__dirname, '../public/noticias_faro.json');
      
      // Verificar se os arquivos existem
      if (fs.existsSync(eventsPath)) {
        const eventsData = fs.readFileSync(eventsPath, 'utf8');
        events = JSON.parse(eventsData);
        console.log(`Carregados ${events.length} eventos`);
      } else {
        console.log(`Arquivo de eventos não encontrado: ${eventsPath}`);
      }
      
      if (fs.existsSync(newsPath)) {
        const newsData = fs.readFileSync(newsPath, 'utf8');
        news = JSON.parse(newsData);
        console.log(`Carregadas ${news.length} notícias`);
      } else {
        console.log(`Arquivo de notícias não encontrado: ${newsPath}`);
      }
    } catch (error) {
      console.error('Erro ao carregar dados:', error);
    }

    // Processar a pergunta
    const answer = processQuestion(question, events, news);
    
    return {
      statusCode: 200,
      headers,
      body: JSON.stringify({ answer })
    };
  } catch (error) {
    console.error('Erro ao processar requisição:', error);
    
    return {
      statusCode: 500,
      headers,
      body: JSON.stringify({ 
        error: 'Erro interno do servidor', 
        details: error.message,
        answer: 'Desculpe, ocorreu um erro ao processar sua solicitação.' 
      })
    };
  }
};

// Função para processar perguntas
function processQuestion(question, events, news) {
  const questionLower = question.toLowerCase();
  console.log('Processando pergunta:', questionLower);
  
  // Perguntas sobre eventos
  if (questionLower.includes('evento') || 
      questionLower.includes('acontecendo') || 
      questionLower.includes('programação')) {
    
    if (events.length === 0) {
      return 'Desculpe, não tenho informações sobre eventos no momento.';
    }
    
    // Listar eventos
    if (questionLower.includes('listar') || 
        questionLower.includes('quais') || 
        questionLower.includes('mostrar')) {
      
      const eventList = events
        .slice(0, 5)
        .map(event => `- ${event.title} (${event.date}${event.time ? ', ' + event.time : ''} em ${event.location})`)
        .join('\n');
      
      return `Aqui estão os próximos eventos em Faro:\n\n${eventList}\n\nEstes são apenas os 5 próximos eventos. Há um total de ${events.length} eventos disponíveis.`;
    }
    
    // Buscar por categoria
    const categoryMatch = questionLower.match(/categoria\s+(\w+)/) || questionLower.match(/eventos\s+de\s+(\w+)/);
    if (categoryMatch) {
      const category = categoryMatch[1];
      const filteredEvents = events.filter(event => 
        event.category && event.category.toLowerCase().includes(category)
      );
      
      if (filteredEvents.length > 0) {
        const eventList = filteredEvents
          .slice(0, 3)
          .map(event => `- ${event.title} (${event.date})`)
          .join('\n');
        
        return `Encontrei ${filteredEvents.length} eventos na categoria '${category}':\n\n${eventList}`;
      } else {
        return `Não encontrei eventos na categoria '${category}'.`;
      }
    }
    
    // Buscar por local
    if (questionLower.includes('local') || questionLower.includes('onde')) {
      const locations = [...new Set(events
        .filter(event => event.location)
        .map(event => event.location)
      )];
      
      return `Os eventos acontecem nos seguintes locais:\n\n${locations.join('\n')}`;
    }
    
    // Buscar por data
    if (questionLower.includes('quando') || questionLower.includes('data')) {
      // Ordenar eventos por data
      const sortedEvents = [...events].sort((a, b) => 
        new Date(a.date || '9999-99-99').getTime() - new Date(b.date || '9999-99-99').getTime()
      );
      
      const eventList = sortedEvents
        .slice(0, 3)
        .map(event => `- ${event.title}: ${event.date}${event.time ? ', ' + event.time : ''}`)
        .join('\n');
      
      return `Os próximos eventos são:\n\n${eventList}`;
    }
    
    // Resposta genérica sobre eventos
    return `Temos ${events.length} eventos disponíveis. Você pode perguntar sobre eventos por categoria, local ou data.`;
  }
  
  // Perguntas sobre notícias
  if (questionLower.includes('notícia') || 
      questionLower.includes('jornal') || 
      questionLower.includes('informação')) {
    
    if (news.length === 0) {
      return 'Desculpe, não tenho informações sobre notícias no momento.';
    }
    
    // Listar notícias recentes
    if (questionLower.includes('recentes') || 
        questionLower.includes('últimas') || 
        questionLower.includes('atuais')) {
      
      const newsList = news
        .slice(0, 5)
        .map(item => `- ${item.title} (${item.source && item.source.name ? item.source.name : 'Fonte desconhecida'})`)
        .join('\n');
      
      return `Aqui estão as notícias mais recentes sobre Faro:\n\n${newsList}`;
    }
    
    // Buscar por fonte
    if (questionLower.includes('fonte') || questionLower.includes('jornal')) {
      const sources = [...new Set(news
        .filter(item => item.source && item.source.name)
        .map(item => item.source.name)
      )];
      
      return `As notícias são de diversas fontes, incluindo:\n\n${sources.join('\n')}`;
    }
    
    // Buscar por tema
    const keywords = ['cultura', 'esporte', 'política', 'economia', 'turismo', 'farense'];
    for (const keyword of keywords) {
      if (questionLower.includes(keyword)) {
        const filteredNews = news.filter(item => 
          (item.title && item.title.toLowerCase().includes(keyword)) || 
          (item.description && item.description.toLowerCase().includes(keyword))
        );
        
        if (filteredNews.length > 0) {
          const newsList = filteredNews
            .slice(0, 3)
            .map(item => `- ${item.title}`)
            .join('\n');
          
          return `Encontrei ${filteredNews.length} notícias sobre '${keyword}':\n\n${newsList}`;
        } else {
          return `Não encontrei notícias específicas sobre '${keyword}'.`;
        }
      }
    }
    
    // Resposta genérica sobre notícias
    return `Tenho ${news.length} notícias disponíveis sobre Faro e região. Você pode perguntar sobre notícias recentes ou por temas específicos como cultura, esporte, política, etc.`;
  }
  
  // Perguntas sobre o site/aplicação
  if (questionLower.includes('site') || 
      questionLower.includes('aplicação') || 
      questionLower.includes('ligafaro') || 
      questionLower.includes('plataforma')) {
    
    if (questionLower.includes('o que é') || questionLower.includes('sobre o site')) {
      return 'O LigaFaro é uma plataforma que conecta os cidadãos de Faro a informações locais, eventos, notícias e serviços comunitários. Nossa missão é fortalecer a comunidade local através da informação e participação.';
    }
    
    if (questionLower.includes('funcionalidades') || questionLower.includes('o que posso fazer')) {
      return 'No LigaFaro você pode:\n\n- Consultar eventos locais\n- Ler notícias sobre Faro e região\n- Participar do fórum comunitário\n- Acessar o marketplace local\n- Encontrar oportunidades de voluntariado';
    }
    
    if (questionLower.includes('contato') || questionLower.includes('suporte')) {
      return 'Para entrar em contato conosco, envie um email para contato@ligafaro.pt ou utilize o formulário de contato disponível no site.';
    }
    
    return 'O LigaFaro é sua plataforma comunitária para Faro. Posso ajudar com informações sobre eventos, notícias locais e funcionalidades do site.';
  }
  
  // Verificar se é uma pergunta sobre o tempo
  if (questionLower.includes('tempo') || questionLower.includes('clima')) {
    // Obter a data e hora atual em Faro (Europe/Lisbon)
    const agora = new Date();
    const options = {
      timeZone: 'Europe/Lisbon',
      weekday: 'long',
      year: 'numeric',
      month: 'long',
      day: 'numeric',
      hour: '2-digit',
      minute: '2-digit'
    };
    const dataHoraFaro = agora.toLocaleString('pt-BR', options);
    
    return `A temperatura atual em Faro é de aproximadamente 22°C, com céu parcialmente nublado. Agora são ${dataHoraFaro} em Faro, Portugal.`;
  }
  
  // Verificar se é uma pergunta sobre data ou hora
  if (questionLower.includes('que dia é hoje') ||
      questionLower.includes('data') ||
      questionLower.includes('hora') ||
      questionLower.includes('horas')) {
    
    const agora = new Date();
    const options = {
      timeZone: 'Europe/Lisbon',
      weekday: 'long',
      year: 'numeric',
      month: 'long',
      day: 'numeric',
      hour: '2-digit',
      minute: '2-digit'
    };
    
    if (questionLower.includes('hora') || questionLower.includes('horas')) {
      const horaFaro = agora.toLocaleTimeString('pt-BR', {
        timeZone: 'Europe/Lisbon',
        hour: '2-digit',
        minute: '2-digit'
      });
      return `Agora são ${horaFaro} em Faro, Portugal.`;
    } else {
      const dataFaro = agora.toLocaleDateString('pt-BR', {
        timeZone: 'Europe/Lisbon',
        weekday: 'long',
        year: 'numeric',
        month: 'long',
        day: 'numeric'
      });
      return `Hoje é ${dataFaro} em Faro, Portugal.`;
    }
  }
  
  // Perguntas sobre voluntariado
  if (questionLower.includes('voluntariado') ||
      questionLower.includes('voluntário') ||
      questionLower.includes('ajudar') ||
      questionLower.includes('contribuir')) {
    
    if (questionLower.includes('oportunidades') || questionLower.includes('como posso')) {
      return `Temos várias oportunidades de voluntariado em Faro, incluindo:
      
  - Limpeza da Praia de Faro (Ambiente Faro)
  - Festival de Artes para Crianças (Associação Cultural de Faro)
  - Apoio ao Centro de Idosos (Centro Social Sénior de Faro)
  - Distribuição de Alimentos (Banco Alimentar de Faro)
  
  Você pode encontrar mais detalhes e se inscrever na seção de Voluntariado do nosso site.`;
    }
    
    if (questionLower.includes('benefícios') || questionLower.includes('por que')) {
      return `O voluntariado traz diversos benefícios:
  
  1. Contribui para o bem-estar da comunidade
  2. Permite conhecer novas pessoas e expandir sua rede de contatos
  3. Desenvolve novas habilidades e experiências
  4. Proporciona satisfação pessoal ao ajudar os outros
  
  Em Faro, valorizamos muito a participação dos voluntários em diversas iniciativas comunitárias.`;
    }
    
    return `O LigaFaro oferece diversas oportunidades de voluntariado em áreas como ambiente, cultura, apoio social e distribuição de alimentos. Visite a seção de Voluntariado no nosso site para conhecer todas as oportunidades disponíveis e como se inscrever.`;
  }
  
  // Perguntas sobre comunidade
  if (questionLower.includes('comunidade') ||
      questionLower.includes('membros') ||
      questionLower.includes('pessoas') ||
      questionLower.includes('conectar')) {
    
    if (questionLower.includes('como participar') || questionLower.includes('como fazer parte')) {
      return `Para participar da comunidade LigaFaro, você pode:
  
  1. Criar um perfil na nossa plataforma
  2. Participar dos eventos locais
  3. Contribuir no fórum comunitário
  4. Inscrever-se em oportunidades de voluntariado
  5. Conectar-se com outros membros através da seção Comunidade`;
    }
    
    if (questionLower.includes('benefícios') || questionLower.includes('vantagens')) {
      return `Fazer parte da comunidade LigaFaro oferece várias vantagens:
  
  1. Acesso a informações locais relevantes
  2. Oportunidades de networking com outros moradores
  3. Participação em decisões comunitárias
  4. Descontos em eventos e serviços locais
  5. Sentimento de pertencimento e conexão com Faro`;
    }
    
    return `A comunidade LigaFaro é formada por diversos membros locais, incluindo líderes comunitários, donos de negócios, artistas, estudantes, professores e muito mais. Na seção Comunidade do nosso site, você pode conhecer e se conectar com outros membros que compartilham interesses semelhantes.`;
  }
  
  // Resposta padrão
  return `Posso ajudar com informações sobre eventos em Faro, notícias locais, voluntariado, comunidade ou sobre o LigaFaro. Temos ${events.length} eventos e ${news.length} notícias disponíveis. Também posso informar sobre o tempo atual e a data em Faro.`;
  }

// Exportar a função processQuestion e o handler para uso em outros módulos
export { processQuestion, handler };

File: src/services/python/chatbot.py
import json
import sys
import os
import re

def process_question(question):
    """Processa a pergunta e retorna uma resposta"""
    
    # Carregar dados de eventos e notícias
    events = []
    news = []
    
    # Caminho base para os arquivos
    base_dir = os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))
    
    # Carregar eventos
    try:
        events_path = os.path.join(base_dir, 'public', 'events_data.json')
        print(f"Tentando carregar eventos de: {events_path}", file=sys.stderr)
        
        if os.path.exists(events_path):
            with open(events_path, 'r', encoding='utf-8') as f:
                events = json.load(f)
                print(f"Carregados {len(events)} eventos", file=sys.stderr)
        else:
            print(f"Arquivo de eventos não encontrado em: {events_path}", file=sys.stderr)
    except Exception as e:
        print(f"Erro ao carregar eventos: {str(e)}", file=sys.stderr)
    
    # Carregar notícias
    try:
        news_path = os.path.join(base_dir, 'public', 'noticias_faro.json')
        print(f"Tentando carregar notícias de: {news_path}", file=sys.stderr)
        
        if os.path.exists(news_path):
            with open(news_path, 'r', encoding='utf-8') as f:
                news = json.load(f)
                print(f"Carregadas {len(news)} notícias", file=sys.stderr)
        else:
            print(f"Arquivo de notícias não encontrado em: {news_path}", file=sys.stderr)
    except Exception as e:
        print(f"Erro ao carregar notícias: {str(e)}", file=sys.stderr)
    
    # Lógica para responder perguntas
    question_lower = question.lower()
    print(f"Processando pergunta: {question_lower}", file=sys.stderr)
    
    # Perguntas sobre eventos
    if any(word in question_lower for word in ['evento', 'acontecendo', 'programação', 'agenda']):
        if len(events) == 0:
            return "Não tenho informações sobre eventos no momento."
        
        # Listar eventos
        if any(phrase in question_lower for phrase in ['listar', 'quais', 'mostrar']):
            event_list = "\n".join([f"- {event['title']} ({event['date']}{', ' + event['time'] if 'time' in event and event['time'] else ''} em {event['location']})" for event in events[:5]])
            return f"Aqui estão os próximos eventos em Faro:\n\n{event_list}\n\nEstes são apenas os 5 próximos eventos. Há um total de {len(events)} eventos disponíveis."
        
        # Buscar por categoria
        category_match = re.search(r'categoria\s+(\w+)', question_lower) or re.search(r'eventos\s+de\s+(\w+)', question_lower)
        if category_match:
            category = category_match.group(1)
            filtered_events = [event for event in events if 'category' in event and category.lower() in event['category'].lower()]
            
            if filtered_events:
                event_list = "\n".join([f"- {event['title']} ({event['date']})" for event in filtered_events[:3]])
                return f"Encontrei {len(filtered_events)} eventos na categoria '{category}':\n\n{event_list}"
            else:
                return f"Não encontrei eventos na categoria '{category}'."
        
        # Buscar por local
        if any(word in question_lower for word in ['local', 'onde']):
            locations = list(set([event['location'] for event in events if 'location' in event]))
            return f"Os eventos acontecem nos seguintes locais:\n\n{chr(10).join(locations)}"
        
        # Buscar por data
        if any(word in question_lower for word in ['quando', 'data']):
            # Ordenar eventos por data
            sorted_events = sorted(events, key=lambda x: x['date'] if 'date' in x else '9999-99-99')
            event_list = "\n".join([f"- {event['title']}: {event['date']}{', ' + event['time'] if 'time' in event and event['time'] else ''}" for event in sorted_events[:3]])
            return f"Os próximos eventos são:\n\n{event_list}"
        
        # Resposta genérica sobre eventos
        return f"Temos {len(events)} eventos disponíveis. Você pode perguntar sobre eventos por categoria, local ou data."
    
    # Perguntas sobre notícias
    if any(word in question_lower for word in ['notícia', 'jornal', 'informação', 'novidade']):
        if len(news) == 0:
            return "Não tenho informações sobre notícias no momento."
        
        # Listar notícias recentes
        if any(phrase in question_lower for phrase in ['recentes', 'últimas', 'atuais']):
            news_list = "\n".join([f"- {item['title']} ({item['source']['name'] if 'source' in item and 'name' in item['source'] else 'Fonte desconhecida'})" for item in news[:5]])
            return f"Aqui estão as notícias mais recentes sobre Faro:\n\n{news_list}"
        
        # Buscar por fonte
        if any(word in question_lower for word in ['fonte', 'jornal']):
            sources = list(set([item['source']['name'] for item in news if 'source' in item and 'name' in item['source']]))
            return f"As notícias são de diversas fontes, incluindo:\n\n{chr(10).join(sources)}"
        
        # Buscar por tema
        keywords = ['cultura', 'esporte', 'política', 'economia', 'turismo', 'farense']
        for keyword in keywords:
            if keyword in question_lower:
                filtered_news = [
                    item for item in news 
                    if (keyword in item['title'].lower()) or 
                       ('description' in item and item['description'] and keyword in item['description'].lower())
                ]
                
                if filtered_news:
                    news_list = "\n".join([f"- {item['title']}" for item in filtered_news[:3]])
                    return f"Encontrei {len(filtered_news)} notícias sobre '{keyword}':\n\n{news_list}"
                else:
                    return f"Não encontrei notícias específicas sobre '{keyword}'."
        
        # Resposta genérica sobre notícias
        return f"Tenho {len(news)} notícias disponíveis sobre Faro e região. Você pode perguntar sobre notícias recentes ou por temas específicos como cultura, esporte, política, etc."
    
    # Perguntas sobre o site/aplicação
    if any(word in question_lower for word in ['site', 'aplicação', 'ligafaro', 'plataforma']):
        if 'o que é' in question_lower or 'sobre o site' in question_lower:
            return 'O LigaFaro é uma plataforma que conecta os cidadãos de Faro a informações locais, eventos, notícias e serviços comunitários. Nossa missão é fortalecer a comunidade local através da informação e participação.'
        
        if 'funcionalidades' in question_lower or 'o que posso fazer' in question_lower:
            return 'No LigaFaro você pode:\n\n- Consultar eventos locais\n- Ler notícias sobre Faro e região\n- Participar do fórum comunitário\n- Acessar o marketplace local\n- Encontrar oportunidades de voluntariado'
        
        if 'contato' in question_lower or 'suporte' in question_lower:
            return 'Para entrar em contato conosco, envie um email para contato@ligafaro.pt ou utilize o formulário de contato disponível no site.'
        
        return 'O LigaFaro é sua plataforma comunitária para Faro. Posso ajudar com informações sobre eventos, notícias locais e funcionalidades do site.'
    
    # Verificar se é uma pergunta sobre o tempo
    if any(word in question_lower for word in ['tempo', 'clima']):
        from datetime import datetime
        import pytz
        
        # Obter a data e hora atual em Faro (Europe/Lisbon)
        faro_tz = pytz.timezone('Europe/Lisbon')
        agora = datetime.now(faro_tz)
        data_hora_faro = agora.strftime('%A, %d de %B de %Y às %H:%M')
        
        return f'A temperatura atual em Faro é de aproximadamente 22°C, com céu parcialmente nublado. Agora são {data_hora_faro} em Faro, Portugal.'
    
    # Verificar se é uma pergunta sobre data ou hora
    if 'que dia é hoje' in question_lower or 'data' in question_lower or 'hora' in question_lower or 'horas' in question_lower:
        from datetime import datetime
        import pytz
        
        # Obter a data e hora atual em Faro (Europe/Lisbon)
        faro_tz = pytz.timezone('Europe/Lisbon')
        agora = datetime.now(faro_tz)
        
        if 'hora' in question_lower or 'horas' in question_lower:
            hora_faro = agora.strftime('%H:%M')
            return f'Agora são {hora_faro} em Faro, Portugal.'
        else:
            data_faro = agora.strftime('%A, %d de %B de %Y')
            return f'Hoje é {data_faro} em Faro, Portugal.'
    
    # Perguntas sobre voluntariado
    if any(word in question_lower for word in ['voluntariado', 'voluntário', 'ajudar', 'contribuir']):
        
        if any(phrase in question_lower for phrase in ['oportunidades', 'como posso']):
            return """Temos várias oportunidades de voluntariado em Faro, incluindo:
            
- Limpeza da Praia de Faro (Ambiente Faro)
- Festival de Artes para Crianças (Associação Cultural de Faro)
- Apoio ao Centro de Idosos (Centro Social Sénior de Faro)
- Distribuição de Alimentos (Banco Alimentar de Faro)

Você pode encontrar mais detalhes e se inscrever na seção de Voluntariado do nosso site."""
        
        if any(phrase in question_lower for phrase in ['benefícios', 'por que']):
            return """O voluntariado traz diversos benefícios:

1. Contribui para o bem-estar da comunidade
2. Permite conhecer novas pessoas e expandir sua rede de contatos
3. Desenvolve novas habilidades e experiências
4. Proporciona satisfação pessoal ao ajudar os outros

Em Faro, valorizamos muito a participação dos voluntários em diversas iniciativas comunitárias."""
        
        return "O LigaFaro oferece diversas oportunidades de voluntariado em áreas como ambiente, cultura, apoio social e distribuição de alimentos. Visite a seção de Voluntariado no nosso site para conhecer todas as oportunidades disponíveis e como se inscrever."
    
    # Perguntas sobre comunidade
    if any(word in question_lower for word in ['comunidade', 'membros', 'pessoas', 'conectar']):
        
        if any(phrase in question_lower for phrase in ['como participar', 'como fazer parte']):
            return """Para participar da comunidade LigaFaro, você pode:

1. Criar um perfil na nossa plataforma
2. Participar dos eventos locais
3. Contribuir no fórum comunitário
4. Inscrever-se em oportunidades de voluntariado
5. Conectar-se com outros membros através da seção Comunidade"""
        
        if any(phrase in question_lower for phrase in ['benefícios', 'vantagens']):
            return """Fazer parte da comunidade LigaFaro oferece várias vantagens:

1. Acesso a informações locais relevantes
2. Oportunidades de networking com outros moradores
3. Participação em decisões comunitárias
4. Descontos em eventos e serviços locais
5. Sentimento de pertencimento e conexão com Faro"""
        
        return "A comunidade LigaFaro é formada por diversos membros locais, incluindo líderes comunitários, donos de negócios, artistas, estudantes, professores e muito mais. Na seção Comunidade do nosso site, você pode conhecer e se conectar com outros membros que compartilham interesses semelhantes."
    
    # Resposta padrão
    return f"Posso ajudar com informações sobre eventos em Faro, notícias locais, voluntariado, comunidade ou sobre o LigaFaro. Temos {len(events)} eventos e {len(news)} notícias disponíveis. Também posso informar sobre o tempo atual e a data em Faro."

if __name__ == "__main__":
    # Receber a pergunta como argumento
    if len(sys.argv) > 1:
        question = sys.argv[1]
        response = process_question(question)
        print(json.dumps({"answer": response}))
    else:
        print(json.dumps({"error": "Nenhuma pergunta fornecida"}))